### ISSUE TITLE
inference time question

### ISSUE THREAD

**Issue by chentairan** [zylo117/Yet-Another-EfficientDet-Pytorch#23]
Hi,

Your implementation has greatly improved the speed. But it seems still slow compared with yolov3 since EfficientDet-D0 has a substantial parameter reduction. 

Is postprocess restricting the running time?
Do you test the average cost time of YOLOv3 running on the 2080Ti ? 

Thanks,
Tairan Chen

**Comment by zylo117** [comment]
No, I haven't test yolo v3 inference time on 2080, because yolo v3' speed can't be replicated on any framework other than Darknet. And installing Darknet is also painful...

Yes, the post process is relatively slow, compared to the model inference, it includes nms, coordinatea affine and GPU tensor to CPU numpy array. I do this deliberately, just want the test to be practical, to show people how fast you can get when you use this model, instead of cutting everything off to get a astonishing sota speed. 

Lots of people do that on the paper, they test only the model part, from the input to the regression and it's done. It's not even the final result.

If you guys are interested about the speed without postprocessing, I will test it later.

**Comment by zylo117** [comment]
You're right, I totally underestimate its speed. I tried running with batch_size 32, D0 I got 163 FPS now, much faster than the paper's

**Comment by chentairan** [comment]
It may be the limitation of bandwidth rather than the limitation of the network itself. You use batch size 32 for testing, which increases the GPU memory usage, in return for speed improvements. Maybe setting the batch size to 1 on a GPU with poor performance such as 1060 will also get the same frame rate I guess.

**Comment by zylo117** [comment]
> It may be the limitation of bandwidth rather than the limitation of the network itself. You use batch size 32 for testing, which increases the GPU memory usage, in return for speed improvements. Maybe setting the batch size to 1 on a GPU with poor performance such as 1060 will also get the same frame rate I guess.

Indeed, the community has get the same speed as 2080 on a 1070ti.
But I don't think is the bandwidth but the cuda processors, When running D0 with bs1, the cuda processors are not fully utilized, so bs32 is much faster. However, D1/D2 don't get such speedup, because the most of the cuda processors have already been utilized.

**Comment by gengyanlei** [comment]
@zylo117 
Hello, author. When I test img (1920 * 1080) with D0:
inferring image for 10 times ...
0.1142204761505127 seconds, 8.75499764755193 FPS, @batch_size 1.
This should include the time consumption of pre-processing and post-processing, which is not the same as the time consumption you provided: 1 / 36.20(FPS)?

**Comment by zylo117** [comment]
> @zylo117
> Hello, author. When I test img (1920 * 1080) with D0:
> inferring image for 10 times ...
> 0.1142204761505127 seconds, 8.75499764755193 FPS, @batch_size 1.
> This should include the time consumption of pre-processing and post-processing, which is not the same as the time consumption you provided: 1 / 36.20(FPS)?

set force_input_size to None, not 1920

**Comment by gengyanlei** [comment]
@zylo117 
Sorry, I didn't notice this detail. 
tks.

**Comment by glenn-jocher** [comment]
@chentairan you might want to see https://github.com/ultralytics/yolov3 for very fast inference at about 43mAP:
```Speed: 11.4/2.1/13.5 ms inference/NMS/total per 608x608 image at batch-size 1```

**Comment by zylo117** [comment]
> @chentairan you might want to see https://github.com/ultralytics/yolov3 for very fast inference at about 43mAP:
> `Speed: 11.4/2.1/13.5 ms inference/NMS/total per 608x608 image at batch-size 1`

Really? That fast? But the repo says it's 19ms total at batch_size 16, how does the speed of batch_size 1 be faster than batch_size 16?

**Comment by glenn-jocher** [comment]
@zylo117 ah, yes sorry for the confusion. The 19ms speed is for multi-scale testing at 640 (45.6 mAP). To be exact:

multi-scale testing at img-size 604, batch-size 16, 45.6 mAP:
`Speed: 17.5/2.3/19.9 ms inference/NMS/total per 640x640 image at batch-size 16`

single-scale testing at batch-size 1, 43.1 mAP:
`Speed: 11.4/2.2/13.6 ms inference/NMS/total per 608x608 image at batch-size 1`

This is for FP32 inference. I tried FP16 with .half() but ironically did not see any speedup (assuming same batch-size). The iDetection iOS app does use discretization however, it runs yolov3-spp at img-size 320 about 30 FPS at FP8, but these coreml models were discretized with coremltools, not pytorch.

**Comment by glenn-jocher** [comment]
@zylo117 actually, perhaps we could update the iDetection app with your efficientdet models. That would be exciting.

At the moment, it has realtime comparison of yolov3-spp-ultralytics and yolov3-tiny, and you can switch batch and forth instantly with a selector button:

<img src="https://user-images.githubusercontent.com/26833433/79385746-f1755880-7f1d-11ea-9a99-f70bde9afb04.jpeg" width="400" />

**Comment by zylo117** [comment]
> @zylo117 ah, yes sorry for the confusion. The 19ms speed is for multi-scale testing at 640 (45.6 mAP). Multi-scale runs at 3 different scales, so 3x the FLOPS. To be exact:
> 
> multi-scale testing at img-size 640, batch-size 16, 45.6 mAP:
> `Speed: 17.5/2.3/19.9 ms inference/NMS/total per 640x640 image at batch-size 16`
> 
> single-scale testing at batch-size 1, 43.1 mAP:
> `Speed: 11.4/2.2/13.6 ms inference/NMS/total per 608x608 image at batch-size 1`
> 
> This is for FP32 inference. I tried FP16 with .half() but ironically did not see any speedup (assuming same batch-size). The iDetection iOS app does use discretization however, it runs yolov3-spp at img-size 320 about 30 FPS at FP8, but these coreml models were discretized with coremltools, not pytorch.

Thank you for sharing. And may I ask would fp16 and fp8 cause precision loss? I tried evaluate effdet in fp16, but the result is terrible. I think extra finetune is needed, right?

**Comment by glenn-jocher** [comment]
Yes quantization will cause some mAP decrease, but from fp32 to fp16 the difference should be pretty small, almost unnoticeable. It’s possible you had a bug?
